crypto/rsa/rsa_ossl.c

```diff
@@ -747,6 +818,67 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
     if (!BN_add(r0, r1, m1))
         goto err;
 
+    /* add m_i to m in multi-prime case */
+    if (primes > 0) {
+        BIGNUM *pr2 = BN_new();
+
+        if (pr2 == NULL)
+            goto err;
+
+        for (i = 0; i < primes; i++) {
+            pinfo = sk_RSA_PRIME_INFO_value(rsa->prime_infos, i);
+            if (!BN_sub(r1, m[i], r0)) {
+                BN_free(pr2);
+                goto err;
+            }
+
+            /*
+             * This will help stop the size of r0 increasing, which does affect
+             * the multiply if it optimised for a power of 2 size
+             */
+            if (BN_is_negative(r1)) {
+                if (!BN_add(r1, r1, pinfo->r)) {
+                    BN_free(pr2);
+                    goto err;
+                }
+            }
+
+            if (!BN_mul(r2, r1, pinfo->t, ctx)) {
+                BN_free(pr2);
+                goto err;
+            }
+
+            BN_with_flags(pr2, r2, BN_FLG_CONSTTIME);
+
+            if (!BN_mod(r1, pr2, pinfo->r, ctx)) {
+                BN_free(pr2);
+                goto err;
+            }
+
+            /*
+             * If r[i - 1] < r[i] it is occasionally possible for the
+             * correction of adding r[i] if r1 is negative above to leave
+             * the result still negative. This can break the private key
+             * operations: the following second correction should *always*
+             * correct this rare occurrence.
+             */
+            if (BN_is_negative(r1))
```

Comment:
I don't follow this comment either. r1 is still negative here, but I fail to see what does it have to do with relation between r[i-1] and r[i].